agent/config/file_watcher_test.go

```diff
@@ -0,0 +1,375 @@
+
+	name := filepath + "/" + randomStr(20)
+	file, err := os.Create(name)
+	require.NoError(t, err)
+	err = file.Close()
+	require.NoError(t, err)
+	err = w.Add(filepath)
+	require.NoError(t, err)
+
+	time.Sleep(w.reconcileTimeout + 50*time.Millisecond)
+	for i := 0; i < 100; i++ {
+		filepathTmp := createTempConfigFile(t, "temp_config2")
+		os.Rename(filepathTmp, name)
+		require.NoError(t, err)
+		require.NoError(t, assertEvent(filepath, watcherCh))
+	}
+}
+
+func TestEventWatcherDirRead(t *testing.T) {
+	watcherCh := make(chan *WatcherEvent)
+	w, err := New(func(event *WatcherEvent) error {
+		fmt.Printf("event for %s\n", event.Filename)
+		watcherCh <- event
+		return nil
+	})
+	defer func() {
+		_ = w.Close()
+	}()
+	w.reconcileTimeout = 20 * time.Millisecond
+	require.NoError(t, err)
+	filepath := createTempConfigDir(t, "temp_config1")
+
+	name := filepath + "/" + randomStr(20)
+	file, err := os.Create(name)
+	require.NoError(t, err)
+	err = file.Close()
+	require.NoError(t, err)
+	err = w.Add(filepath)
+	require.NoError(t, err)
+
+	time.Sleep(w.reconcileTimeout + 50*time.Millisecond)
+	_, err = os.ReadFile(name)
+	require.NoError(t, err)
+	require.Error(t, assertEvent(filepath, watcherCh), "timedout waiting for event")
+}
+
+func TestEventWatcherMoveSoftLink(t *testing.T) {
+	watcherCh := make(chan *WatcherEvent)
+	w, err := New(func(event *WatcherEvent) error {
+		watcherCh <- event
+		return nil
+	})
+	defer func() {
+		_ = w.Close()
+	}()
+	w.reconcileTimeout = 20 * time.Millisecond
+	require.NoError(t, err)
+	filepath := createTempConfigFile(t, "temp_config1")
+	tempDir := createTempConfigDir(t, "temp_dir")
+	name := tempDir + "/" + randomStr(20)
+	err = os.Symlink(filepath, name)
+	require.NoError(t, err)
+	err = w.Add(name)
+	require.Error(t, err, "symbolic link are not supported")
+
+}
+
+func assertEvent(name string, watcherCh chan *WatcherEvent) error {
+	timeout := time.After(2000 * time.Millisecond)
+	select {
+	case ev := <-watcherCh:
+		if ev.Filename != name && !strings.Contains(ev.Filename, name) {
+			return fmt.Errorf("filename do not match %s %s", ev.Filename, name)
+		}
+		return nil
+	case <-timeout:
+		return fmt.Errorf("timedout waiting for event")
+	}
+}
+
+func createTempConfigFile(t *testing.T, filename string) string {
+	file := testutil.TempFile(t, filename)
+	defer func() {
+		err := file.Close()
+		require.NoError(t, err)
+	}()
+	_, err := file.WriteString("test config")
+	require.NoError(t, err)
+	err = file.Sync()
+	require.NoError(t, err)
+	return file.Name()
+}
+
+func createTempConfigDir(t *testing.T, dirname string) string {
+	return testutil.TempDir(t, dirname)
+}
+func randomStr(length int) string {
+	const charset = "abcdefghijklmnopqrstuvwxyz" +
+		"ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+	var seededRand *rand.Rand = rand.New(
+		rand.NewSource(time.Now().UnixNano()))
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = charset[seededRand.Intn(len(charset))]
+	}
+	return string(b)
+}
```

Comment:
Added this helper to generate random strings, but if we already have something similar I could reuse.